servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextProvider.java

```diff
@@ -40,10 +40,30 @@ interface AsyncContextProvider {
     /**
      * Get the current context.
      *
+     * Note that this method is for getting the {@link ContextMap} for use by the application code. For saving the
+     * current state for crossing an async boundary see the {@link AsyncContextProvider#saveContext()} method.
+     *
      * @return The current context.
      */
     ContextMap context();
 
+    /**
+     * Save existing context in preparation for an asynchronous thread jump.
+     *
+     * Note that this can do more than just package up the ServiceTalk {@link AsyncContext} and could be enhanced or
+     * wrapped to bundle up additional contexts such as the OpenTelemetry or grpc contexts.
+     * @return the saved context state that may be restored later.
+     */
+    ContextMap saveContext();
```

Comment:
In an ideal world we would make the fact that this is a `ContextMap` abstract and have an interface that looks something like this
```
interface SavedContext {
  Scope restoreContext();
}
```
However, our own uses are relatively complex and furthermore people are already writing agent code that make this change difficult. See the DataDog java-agent implementation [here](https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/servicetalk/src/main/java/datadog/trace/instrumentation/servicetalk/ContextPreservingInstrumentation.java#L54-L71) which depends on the saved context state being a `ContextMap`.

Reply:
See https://github.com/bryce-anderson/servicetalk/commit/5806c7841dbdb6af7c4f007e336a0b39e41e5458 for a commit that unwinds current usage of the ContextMap directly. Note that there would still be a bit more to do.